Memory Usage: Add memory transferred between Kokkos Mem Spaces

vlkale opened the pull request:
This PR adds memory transferred between Kokkos Mem Spaces to the Kokkos Tools memory-usage tool library. This is done based on a request in Kokkos Tools Github Issue #50. 

This PR adds functionality to the tool so that the size of data transferred in the deep_copy in a Kokkos application program is accumulated during the execution of a Kokkos program. The deep_copy accumulation is done per Kokkos Memory Space dst->src pair (e.g., OpenMP on host  CUDA on device). Note that the "dst" and "src" Kokkos memory space being the same means that there was a deep_copy in the same memory space. 

When Kokkos is finalized in a Kokkos application program, the finalize callback prints out the deep_copy accumulations per memory space to the file.

vlkale committed:
kp_memory_usage.cpp: put in total memory usage

vlkale committed:
kp_memory_usage.cpp: put in destination and source xferred

vlkale committed:
kp_memory_usage.cpp: put in memory space transferred with from and to

vlkale committed:
kp_memory_usage.cpp: updates to memory transferred

vlkale committed:
kp_memory_usage.cpp: fix llu

vlkale committed:
kp_memory_usage: merge totals with per mem space

vlkale committed:
kp_memory_usage.cpp: put in correct format string

vlkale committed:
kp_memory_usage.cpp: apply clang-format

vlkale committed:
kp_memory_usage.cpp: put in memory usage total

vlkale committed:
kp_memory_usage.cpp: remove totals and apply clang format

vlkale committed:
kp_memory_usage.cpp: num_spaces unsigned int

vlkale committed:
kp_memory_usage.cpp: apply clang-format

vlkale committed:
kp_memory_usage.cpp: remove 4 parameter names from begin_deep_copy since they are not used in the function.

vlkale committed:
kp_memory_usage.cpp: better prints for src and dst of table

vlkale committed:
kp_memory_usage.cpp: fix table headings for prints for src and dst of table

vlkale commented:
Here is output from the stream.cuda in the Kokkos-core benchmarks directory run on Perlmutter on 1 GPU. The number of bytes in the HighWater at the last point in time shown in the first table is equal to the data transferred between host and device. 

```
-----@nid----:~/kks-clean/benchmarks/stream> cat nid001213-1948341-Cuda.memspace_usage
# Space Cuda
# Time(s)  Size(MB)   HighWater(MB)   HighWater-Process(MB)
0.018076 7629.4 7629.4 142.7
0.047229 15258.8 15258.8 143.5
0.071899 22888.2 22888.2 144.0
12.252089 15258.8 22888.2 23032.6
12.313410 7629.4 22888.2 23032.6
12.331277 0.0 22888.2 23032.6
# Data transferred between Kokkos Memory Spaces --- 
# DstSpace     SrcSpace    Data-Transferred(MB)
Cuda Cuda 0.0 
# DstSpace     SrcSpace    Data-Transferred(MB)
Cuda Host 22888.2 
# DstSpace     SrcSpace    Data-Transferred(MB)
Host Cuda 22888.2 
# DstSpace     SrcSpace    Data-Transferred(MB)
Host Host 0.0 

------@nid----:~/kks-clean/benchmarks/stream> cat nid001213-1948341-
nid001213-1948341-Cuda.memspace_usage  nid001213-1948341-Host.memspace_usage  
vkale3@nid001213:~/kks-clean/benchmarks/stream> cat nid001213-1948341-Host.memspace_usage 
# Space Host
# Time(s)  Size(MB)   HighWater(MB)   HighWater-Process(MB)
0.077138 7629.4 7629.4 144.0
0.768111 15258.8 15258.8 7773.6
1.455819 22888.2 22888.2 15403.0
12.218183 15258.8 22888.2 23032.6
12.229001 7629.4 22888.2 23032.6
12.240329 0.0 22888.2 23032.6
# Data transferred between Kokkos Memory Spaces --- 
# DstSpace     SrcSpace    Data-Transferred(MB)
Cuda Cuda 0.0 
# DstSpace     SrcSpace    Data-Transferred(MB)
Cuda Host 22888.2 
# DstSpace     SrcSpace    Data-Transferred(MB)
Host Cuda 22888.2 
# DstSpace     SrcSpace    Data-Transferred(MB)
Host Host 0.0 
```

crtrott reviewed (commented):
Just put it into a separate file otherwise this is good.

dalg24 commented on profiling/memory-usage/kp_memory_usage.cpp:
```diff
@@ -128,6 +142,34 @@ void kokkosp_deallocate_data(const SpaceHandle space, const char* /*label*/,
+  totalMemoryTransferred[space_dst][space_src] += size;
+}
+
+void kokkosp_end_deep_copy() { std::lock_guard<std::mutex> lock(m); }
```
Why do we need to acquire the mutex here?

vlkale replied:
Thanks for pointing it out. This `lock_guard` is not needed and it could degrade performance for this tool.

When I was first outlining this PR, I was thinking I would need to add a code block for tool post-processing at the end of a `Kokkos::deep_copy()` that needs a lock around it.

I will take this out.

dalg24 commented on profiling/memory-usage/kp_memory_usage.cpp:
```diff
@@ -83,6 +87,16 @@ void kokkosp_finalize_library() {
     }
+
+    fprintf(ofile, "# Data transferred between Kokkos Memory Spaces --- \n");
+    for (unsigned int dst = 0; dst < (unsigned int)num_spaces; dst++) {
+      for (unsigned int src = 0; src < (unsigned int)num_spaces; src++) {
```
```suggestion
    for (int dst = 0; dst < num_spaces; dst++) {
      for (int src = 0; src < num_spaces; src++) {
```

vlkale committed:
Update kp_memory_usage.cpp: remove unsigned via dalg24's suggestion

Co-authored-by: Damien L-G <dalg24+github@gmail.com>

vlkale committed:
kp_memory_usage: use separate file for data transfers across Kokkos Memory Spaces

vlkale committed:
kp_memory_usage.cpp: apply clang format

vlkale commented:
I have updated with the requested changes. I did a quick check on my mac laptop and it works on my laptop fine with the change to have a separate file, and for reference, here is the output: 

```
vklap % more /kks/benchmarks/stream/s1ca-13343-Host.memspace_usage 
# Space Host
# Time(s)  Size(MB)   HighWater(MB)   HighWater-Process(MB)
0.000067 762.9 762.9 1856.0
0.084233 1525.9 1525.9 783152.0
0.170814 2288.8 2288.8 1564448.0
3.166797 1525.9 2288.8 2345824.0
3.180026 762.9 2288.8 2345824.0
3.194930 0.0 2288.8 2345824.0

vklap % more /kks/benchmarks/stream/s1ca-13343-Host.memspace_transfers
# Dst Mem Space     Src Mem Space    Total-Data-Transferred(MB)
Host Host 4577.6 
```

(As can be seen, there are now two files rather than one.)

Right now, there seems to be blocker on the CI - I don't know why the CI checks for OpenMP, CUDA and HIP (non-simple builds) are failing now. The problem is arising from Kokkos_Profiling.hpp and is coming from int_for_synchronization_reason(), as can be seen from the logs. I don't think I changed anything in this PR that would impact that? This is a CI check from when it was working previously: 
https://github.com/kokkos/kokkos-tools/actions/runs/10948763633/job/30400520719. The PR is built on top of the latest develop.

vlkale commented:
> Just put it into a separate file otherwise this is good.

This is fixed.

masterleinad commented:
> I don't know why the CI checks for OpenMP, CUDA and HIP (non-simple builds) are failing now

Have you rebased on top of `develop`? That might fix it.

vlkale commented:
> > I don't know why the CI checks for OpenMP, CUDA and HIP (non-simple builds) are failing now
> 
> 
> 
> Have you rebased on top of `develop`? That might fix it.

Thanks for checking this. As mentioned in the last sentence of my previous comment (https://github.com/kokkos/kokkos-tools/pull/272#issuecomment-2519032604), I did rebase on top of `develop` of Kokkos Tools. 

Actually, I think this has to do with Kokkos Tools Issue #275, where there is ultimately an incompatibility with of Kokkos Tools with the current Kokkos version (https://github.com/kokkos/kokkos-tools/issues/275#issuecomment-2422634745). This issue came up in early October, which was after the previous successful run I linked.
I will see if there are quick fixes from that, but given that #275 impacts other PRs, I will advocate for that github issue to be a priority to fix for Kokkos Tools.

vlkale committed:
Merge branch 'memxferred' of https://www.github.com/vlkale/kokkos-tools into memxferred

vlkale committed:
Resolve new merge conflict after pull

vlkale committed:
Apply clang-format-8 to kp_memory_usage.cpp

vlkale committed:
Revert kp_memory_usage.cpp to state from Dec 4, 2024 (commit 2a05419)

vlkale committed:
docs: refine Copilot instructions with sampler details, TPI fences, and repo-specific workflows
